### ISSUE TITLE
Should we assume that certains commands did what we expected them to do?

### ISSUE THREAD

**Issue by cpitclaudel** [rocky/emacs-dbgr#115]
Many debuggers fail to print even basic feedback. GDB's `delete` is one example: it doesn't confirm that a deletion occurred. Same goes for Python's PDB, and probably a bunch of others.

I wonder if an alternate approach could be to assume success of these commands. In practice, this would mean extending some of the commands in cmds.el: for example, cmd-delete would not only issue a "delete breakpoint" command: it would also go ahead an delete that breakpoint.

This is certainly a bit risky: there may be cases where we end up losing track of a breakpoint, for example. Being wrong in a small number of cases, however, still seems better that being consistently wrong :)

And of course, in the long run, we want to become a positive force to convince debugger writers to adopt the same kind of reasonable behavior as @rocky already does in the debuggers he wrote, and be explicit about everything :)

As a use, I think this is the one main thing that I currently see as really limiting; it would be really good to get a good fix for this.

@rocky, what do you think?

**Comment by cpitclaudel** [comment]
A few additional notes:

* This is definitely only a partial solution: for example, typing a command that doesn't produce feedback will still cause us to miss an update.

* We should only do this for idempotent things, ie things that can be safely repeated. So it's ok to do it for deletions (repeating a breakpoint deletion by number isn't bad), but not for, say, `step`, or a breakpoint addition.

**Comment by cpitclaudel** [comment]
Another thought, which could work as an alternate strategy: we could change debugger regexps to also match the corresponding commands. This could handle the "typing in the command buffer" case.

For example, this works for the `delete` command in GDB:

```
(setf (gethash "brkpt-del" realgud:gdb-pat-hash)
      (make-realgud-loc-pat
       :regexp "^\\(?:Deleted breakpoints?\\|^\\s-*delete\\) \\(\\([0-9]+ *\\)+\\)\n"
       :num 1))
```

**Comment by cpitclaudel** [comment]
I'm not sure what I prefer. The first strategy seems more simple to implement, but it might be more brittle / less customizable? If we do pick the second strategy, it could be good to allow lists of patterns instead of single patterns, to make the regexps cleaner.

**Comment by rocky** [comment]
For those debuggers that don't give feedback, realgud can issue another command that queries status and checks output. For example for gdb on a delete,  the status command is `info`:

```
(gdb) break main
Breakpoint 1 at 0x40159c: file ctest.c, line 80.
(gdb) run
Starting program: /home/rocky/c/ctest 

Breakpoint 1, main (argc=1, argv=0x7fffffffde68) at ctest.c:80
80	  i = 1;  num_entries = 2;
(gdb) delete 1
(gdb) info breakpoint 1
No breakpoint or watchpoint matching '1'.
(gdb) 
```
I think somewhere in there is a way to call the backend debugger and not have its output echoed. 

Independently periodically realgud needs to ask for status. For example this occurs after code is running and it wants new information regarding the stack or local variables.

**Comment by cpitclaudel** [comment]
Very interesting, thanks! We would still have to extend the relevant regexp patterns, though, right? For example to parse `No breakpoint or watchpoint matching '1'.`?

Does realgud already issue these `info` commands? Or were you saying that we should make it do so?

**Comment by cpitclaudel** [comment]
Also, I wonder what we should do with debuggers that don't include such `info` commands (PDB doesn't seem to, for example). Maybe for these parsing the input command as well as the output one would work.

**Comment by rocky** [comment]
It's not a bad idea to extend regular expressions include a list of regular expression, since that can simplify the regular expression. 

But I am not sure that is what would be done here. realgud sees you typed a delete command with parameter x. Then it says, oh for this debugger I need to issue "info breakpoint x"  and see what's up with that. So this class of parsing might be a little different.  

The code at 
https://github.com/realgud/realgud/blob/master/realgud/common/buffer/backtrace.el#L119-L136

is along the lines I was thinking. Of course as with a lot of this code, there should be a common routine for doing this kind of stuff. 

Oh, and there probably should be a command you can issue that says resync the information you have about the state of things with what the debugger things. That command might issue a "info breakpoints"  and parse that output. For a many-windows kind of setup this kind of thing happens if nothing else to get local variable and stack frame state.

**Comment by rocky** [comment]
> Also, I wonder what we should do with debuggers that don't include such info commands (PDB doesn't seem to, for example)

In pdb, you issue "b" to get a list. But the general point is of course still valid. 

If the command was issued via realgud then that's simple to handle. If someone enters a command on a debugger command line in the command buffer, then yeah you have to parse input and both eshell and comint (and thus shell) offer an input filter hook which can alert you to a possible command that needs attention.

**Comment by cpitclaudel** [comment]
> If the command was issued via realgud then that's simple to handle. If someone enters a command on a debugger command line in the command buffer, then yeah you have to parse input and both eshell and comint (and thus shell) offer an input filter hook which can alert you to a possible command that needs attention.

I don't think we need the input filter: we already include the command in the text that we scan for responses (IOW, the regexp that I posted above already works for commands typed by the user in the command buffer)

What about the following approach: I can make a patch to support lists of regexps, and adjust gdb and pdb to use these to parse `delete` commands. These should be fairly small changes.

Later on we can implement a more comprehensive solution, which would include parsing status information and updating our own state accordingly (I need to get a better grasp of the rest of the code before thinking of making such changes :).

Would this work?

**Comment by rocky** [comment]
> Would this work?

Sure.

**Comment by cpitclaudel** [comment]
Turns out it was more invasive than I expected. I'm working on this in the `multiple-patterns-in-loc` branch.  Not ready to merge yet, but in testing phase; help welcome :)

I also refactored `track-from-region` to take a list of filters and apply them one by one.

**Comment by cpitclaudel** [comment]
Question: do you remember why the part of the output matching a location was trimmed out before looking for new or deleted breakpoints?

**Comment by rocky** [comment]
> Question: do you remember why the part of the output matching a location was trimmed out before looking for new or deleted breakpoints?

Yes. Program output can appear there as well.

**Comment by cpitclaudel** [comment]
Ah, tricky. The problem I ran into (#118) is that PDB prints interesting info before its prompts, and we currently discard it :/ Was it to avoid cases in which program output would contain things that look like debugger messages?

**Comment by rocky** [comment]
Yep. That's why something like GDB's mi is cool. Or a socket connection to the debugger, etc.  gdb, bashdb, and other of my debuggers have a `tty` command (although the semantics is a little different).

**Comment by cpitclaudel** [comment]
Makes sense. I wonder what the best option is; my feeling is that missing debugger messages is going to be more common than finding spurious matches in program output, but this is of course just an intuition. Also, nothing really tells us that there won't be program output after a debugger prints location information, right? (although ideally a well written debugger wouldn't do that)

**Comment by ilohmar** [comment]
I really enjoy my first steps with `realgud:pdb`, but not having breakpoint deletion reflected in the UI is very annoying (as is the fact that there can be several breakpoints per line, but that's hardly realgud's fault AFAIK).

How about this (purely from a pdb perspective):
Assume that the command (like clear <n> in my case) was successful (and hence delete the UI breakpoint), unless we can *prove* there was a problem (eg, pdb saying "out of range" or whatever, *and nothing else* in the output).

My feeling is that this way, the UI would be right almost all of the time --- the exception being that the breakpoint could not be deleted and at the same time there was additional output that we could not fully understand.

**Comment by cpitclaudel** [comment]
Hey @ilohmar,

Thanks for the feedback :) I'm working on this in the `multiple-patterns-in-loc` branch; in the meantime, I'd suggest using `D` instead of `X` to delete breakpoints; with `D` pdb prints a message, and everything works fine.

**Comment by ilohmar** [comment]
Oh, actually I have been using "D" all the time!  pdb deletes the breakpoint and prints a message indeed, but the UI still shows the breakpoint --- is that an unknown bug?  If so, it might depend on my setup and I should investigate this.

So, to be clear, you are saying that "D" does not only delete the breakpoint in pdb, but should also currently remove the fringe symbol in the UI, right?

**Comment by cpitclaudel** [comment]
Absolutely; it works fine for me on a clean install of the latest RealGUD. Your help in debugging this would be much appreciated :)

**Comment by ilohmar** [comment]
Hmm, I have a hard time understanding how the code works, but can easily reproduce the problem with a small python file in emacs -q, after `(require 'realgud)`.  From where is the fringe update initiated, anyway?  At first glance it looks like it's only due to `realgud-track-comint-output-filter-hook`, but that is never triggered for the "D" command (or for adding breakpoints) in my tests.

I guess I am too tired right now :)

**Comment by cpitclaudel** [comment]
Let's try to get a repro on my side before debugging it :) Can you confirm your Emacs version? Can you reproduce the problem in emacs -Q?

**Comment by ilohmar** [comment]
From (self-compiled) emacs -Q:
```lisp
(emacs-version)
"GNU Emacs 25.0.94.1 (x86_64-pc-linux-gnu, GTK+ Version 3.20.4)
 of 2016-05-24"
(package-initialize)
nil
(require 'realgud)			;realgud-20160608.1938
realgud
```

Now open the following file `realgud.py`:
```python
def fun():
    x = [2 * i for i in range(34, 55)]
    return x

if __name__ == "__main__":
    fun()
```
do `M-x realgud:pdb`, and run as `python3 -m pdb ~/Documents/prg/realgud.py`.

Set a breakpoint on the `return x` line with `B` in the src buffer, pdb outputs
```
(Pdb) break /home/il/Documents/prg/realgud.py:3
Breakpoint 1 at /home/il/Documents/prg/realgud.py:3
```
Delete the breakpoint with `D` in the src buffer, pdb outputs
```
(Pdb) clear 1
Deleted breakpoint 1 at /home/il/Documents/prg/realgud.py:3
```
But the breakpoint marker in the left fringe of the src buffer is still there.

**Comment by cpitclaudel** [comment]
Thanks, the key missing bit was the fact that you're running Python 3 :) Should be fixed now.

**Comment by ilohmar** [comment]
Awesome, that did not occur to me at all, thanks!
